homeassistant/components/discovergy/config_flow.py

```diff
@@ -0,0 +1,210 @@
+"""Config flow for Discovergy integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import pydiscovergy
+import pydiscovergy.error as discovergyError
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+
+from .const import (
+    APP_NAME,
+    CONF_ACCESS_TOKEN,
+    CONF_ACCESS_TOKEN_SECRET,
+    CONF_CONSUMER_KEY,
+    CONF_CONSUMER_SECRET,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from schema with values provided by the user.
+    """
+    try:
+        discovergy_instance = pydiscovergy.Discovergy(APP_NAME)
+        (access_token, consumer_token) = await discovergy_instance.login(
+            data[CONF_EMAIL], data[CONF_PASSWORD]
+        )
+
+        # store token pairs for later use so we don't need to request new pairs each time
+        data[CONF_CONSUMER_KEY] = consumer_token.key
+        data[CONF_CONSUMER_SECRET] = consumer_token.secret
+        data[CONF_ACCESS_TOKEN] = access_token.token
+        data[CONF_ACCESS_TOKEN_SECRET] = access_token.token_secret
+    except discovergyError.InvalidLogin as err:
+        raise InvalidAuth from err
+    except discovergyError.HTTPError as err:
+        raise CannotConnect(f"Error while communicating with API: {err}") from err
+
+    return {"title": data[CONF_EMAIL], "data": data}
+
+
+def make_schema(user_input: dict[str, Any] = None, step_id: str = "user") -> vol.Schema:
+    """Return the schema filled with user_input defaults."""
+    if step_id == "reauth_confirm":
+        return vol.Schema(
+            {
+                vol.Required(
+                    CONF_PASSWORD,
+                    default=(
+                        user_input[CONF_PASSWORD]
+                        if user_input is not None and CONF_PASSWORD in user_input
+                        else ""
+                    ),
+                ): str,
+            }
+        )
+
+    return vol.Schema(
+        {
+            vol.Required(
+                CONF_EMAIL,
+                default=(
+                    user_input[CONF_EMAIL]
+                    if user_input is not None and CONF_EMAIL in user_input
+                    else ""
+                ),
+            ): str,
+            vol.Required(
+                CONF_PASSWORD,
+                default=(
+                    user_input[CONF_PASSWORD]
+                    if user_input is not None and CONF_PASSWORD in user_input
+                    else ""
+                ),
+            ): str,
+        }
+    )
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```

Comment:
As a general comment, I found the config flow code really hard to follow and I feel like this can be made a lot simpler. Things are jumping around a lot in the code, which is making it more complex and harder to read.

In the end, this is an integration authenticating with a username/password. It shouldn't have to be this large/complex.

Reply:
The config flow has been rewritten :)